### PR TITLE
ROCANA-1952 Add filesystem metrics for Windows

### PR DESCRIPTION
**Pull request by actgardner**

### CODE REVIEW

File: sigar_windows.go [hunk 1]
```diff
@@ -117,8 +118,70 @@ func (self *Cpu) populateFromPdh(values []uint64) {
...
+		chars = make([]uint16, charsLen)
+		if C.GetVolumePathNamesForVolumeNameW((*C.WCHAR)(&volumeName[0]), (*C.WCHAR)(&chars[0]), C.MAX_PATH, &charsLen) == C.TRUE {
+			break
+		}
+		errno := syscall.GetLastError()
+		err, ok := errno.(syscall.Errno)
+		if !ok || err != C.ERROR_MORE_DATA {
+			return "", errno
+		}
+	}
```
Comment (MarkRoddy) [inline sigar_windows.go]: I'm having a lot of trouble understanding the control flow here.  I don't get the purpose of the for loop.  If we're successful or get an error that isn't ERROR_MORE_DATA we return, but if we do get ERROR_MORE_DATA there's no resizing so wouldn't we get stuck in an infinite loop if we get a ERROR_MORE_DATA?  Or is it the case that getting ERROR_MORE_DATA is impossible because we're sizing at MAX_PATH?  But if that's the case the for loop seems unnecessary.

File: sigar_windows.go [hunk 2]
```diff
@@ -117,8 +118,70 @@ func (self *Cpu) populateFromPdh(values []uint64) {
...
+	capacity := len(self.List)
+	if capacity == 0 {
+		capacity = 4
+	}
+	self.List = make([]FileSystem, capacity)
+
+	// Get a handle to iterate over all the attached disks
+	diskHandle := C.FindFirstVolumeW((*C.WCHAR)(&buf[0]), bufSize)
+	if diskHandle == C.HANDLE(uintptr(0xFF)) {
+		return fmt.Errorf("Got bad handle when calling FindFirstVolume")
```
Comment (MarkRoddy) [inline sigar_windows.go]: There's no errorno set in this case?

File: sigar_windows.go (hunk 1 above)
Comment (actgardner) [inline sigar_windows.go]: Yeah, looking at this with fresh eyes, it's super wrong. What we need to do is replace `C.MAX_PATH` in the call to `GetVolumePathNamesForVolumeNameW` with `len(chars)`. Then we'd start with a buffer of size `C.MAX_PATH` and grow it to `charsLen` if the first call fails. 

The convention on MSDN is to wrap these in loops (as above), but it seems a lot better to try twice and fail loudly instead of getting stuck in an infinite loop. I'll rewrite appropriately.

File: sigar_windows.go [hunk 3]
```diff
@@ -117,8 +118,70 @@ func (self *Cpu) populateFromPdh(values []uint64) {
...
+
+		self.List = append(self.List, FileSystem{
+			DirName: volumeMount,
+			DevName: volume,
+		})
+
+		if C.FindNextVolumeW(diskHandle, (*C.WCHAR)(&buf[0]), bufSize) == C.FALSE {
+			errno := syscall.GetLastError()
+			err, _ := errno.(syscall.Errno)
+			if errno == nil || err == C.ERROR_NO_MORE_FILES {
```
Comment (MarkRoddy) [inline sigar_windows.go]: The doc for FindNextVolume says:

> If no matching files can be found, the GetLastError function returns the ERROR_NO_MORE_FILES error code. In that case, close the search with the FindVolumeClose function.

~~Shouldn't we perform ^^ instead of continuing when we get `ERROR_NO_MORE_FILES`?~~

Edit: this is a break not of a continue. nevermind

File: sigar_windows.go [hunk 4]
```diff
@@ -117,8 +118,70 @@ func (self *Cpu) populateFromPdh(values []uint64) {
...
+		if err != nil {
+			return fmt.Errorf("Error getting volume mountpoints: %v", err)
+		}
+
+		self.List = append(self.List, FileSystem{
+			DirName: volumeMount,
+			DevName: volume,
+		})
+
+		if C.FindNextVolumeW(diskHandle, (*C.WCHAR)(&buf[0]), bufSize) == C.FALSE {
```
Comment (MarkRoddy) [inline sigar_windows.go]: I had to reread this function a bunch of times, and I really didn't understand what was going on until I read through the docs for `FindFirstVolume` and `FindNextVolume`.  I think these are kind of baroque and made this function hard to follow.  Have we thought about wrapping them in some struct that we can use some what like an iterator?

File: volume_windows.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+	handle     C.HANDLE
+	buffer     []uint16
+	bufferSize C.DWORD
+	volume     FileSystem
+	err        error
+	hasNext    bool
+}
+
+// Get the first mount point for the volume
+func getVolumePathName(wideName []uint16, bufferSize uint16) (string, error, uint16) {
```
Comment (MarkRoddy) [inline volume_windows.go:24]: What is the third value being returned?

File: sigar_windows.go [hunk 6]
```diff
@@ -117,8 +117,27 @@ func (self *Cpu) populateFromPdh(values []uint64) {
...
+
+	iter, err := NewWindowsVolumeIterator()
+	if err != nil {
+		return err
+	}
+
+	for iter.Next() {
+		volume := iter.Volume()
+		self.List = append(self.List, volume)
+	}
```
Comment (MarkRoddy) [inline sigar_windows.go:137]: Very Nice

File: volume_windows.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	/* Because of the weird structure of `FindFirstVolume`,
+	   we have a volume buffered before the first `Next()`. */
+	if !self.hasNext {
+		return false
+	}
+
+	// Convert the currently buffered event to a Go struct
+	volume := syscall.UTF16ToString(self.buffer)
+
+	// Get the first mount point for each volume
```
Comment (MarkRoddy) [inline volume_windows.go]: should this read "Get the first mount point for the next volume"?

File: volume_windows.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+	volumeMount, err, newSize := getVolumePathName(self.buffer, C.MAX_PATH)
+	if newSize > 0 {
+		// Retry once if we get an error about buffers being too small
+		volumeMount, err, _ = getVolumePathName(self.buffer, newSize)
```
Comment (MarkRoddy) [inline volume_windows.go:88]: Any chance we could lose the error from the previous call to `getVolumePathName()`?

**Comment by MarkRoddy** [comment]
A couple of questions, but this is looking 1000x better!

File: volume_windows.go (hunk 8 above)
Comment (actgardner) [inline volume_windows.go:88]: If `newSize > 0` we know the last error was requesting a bigger buffer. So there's no risk of losing the last value of `err`.

Comment (MarkRoddy) [inline volume_windows.go:88]: Not necessary to block on, but it's worth some though on a way to make that behavioral dependency explicit rather than implicit.

**Comment by actgardner** [comment]
Added some docs to clarify the questions

**Comment by MarkRoddy** [comment]
lgtm
